src/postgres/src/backend/storage/ipc/procarray.c

```diff
@@ -2794,6 +2810,39 @@ MinimumActiveBackends(int min)
 	return count >= min;
 }
 
+void insertNode(PgProcAuhNode **head, PGProcAUHEntryList data) {
+    PgProcAuhNode *newNode = malloc(sizeof(PgProcAuhNode));
+    newNode->data = data;
+    newNode->next = *head;
+    *head = newNode;
+}
+void freeLinkedList(PgProcAuhNode *head) {
+    while (head != NULL) {
+        PgProcAuhNode *temp = head;
+        head = head->next;
+        free(temp);
+    }
+}
+
+PgProcAuhNode* pg_collect_samples_proc()
+{
+	ProcArrayStruct *arrayP = procArray;
+	LWLockAcquire(ProcArrayLock, LW_SHARED);
+	PgProcAuhNode *head = NULL;
+	for (int i = 0; i < procArray->numProcs; i++)
+	{
+		int	pgprocno = arrayP->pgprocnos[i];
+		PGPROC *proc  = &allProcs[pgprocno];
+		if(proc != NULL && proc->pid != 0 && proc->wait_event_info!=0)
+		{
+			PGProcAUHEntryList entry = proc_getter(proc, procArray->numProcs);
```

Comment:
We don't want numProcs, but the final size of the linked list which won't be equal to numprocs always